### PR TITLE
LPS-124112 Create ERC endpoints for KBArticle and KBFolder. Modify rest builder.

### PR DESCRIPTION
**Pull request by LuismiBarcos**
I create this PR for:
* Review the changes on the rest builder to add a test for upsert in ERC endpoints
* Discuss if we should change the implementation of the ERC in the Account module.
* I can't execute `gw buildREST` in the `modules/dxp/apps` or any API inside this path. Every time I execute the command, I receive a `BUILD SUCCESSFUL` but the time of the execution is suspicious, it should take long.

**About the implementation of the ERC in the Account module:**
I`ve seen in the file [AccountResourceImpl](https://github.com/liferay/liferay-portal/blob/0d801ac0a9a598c2950f8c594de7e45826a413fa/modules/apps/account/account-rest-impl/src/main/java/com/liferay/account/rest/internal/resource/v1_0/AccountResourceImpl.java#L225-L249) that the implementation of the ERC for PUT only updates the entities. So, it doesn't support UPSERT like we are doing for headless-delivery.

Is this on purpose? Should we, as an infrastructure team, change it?

**About the changes in the rest builder:**
Thanks to @javierdearcos for helping me here. I added a particularity which is a protected method to create new entities. This protected method just returns a `randomEntity`. The purpose of this is to provide a way to change the creation because in some special cases, it is necessary to set an Id. For example, for the creation of a WikiPage by ERC, we need to set the WikiNodeId to specify where the WikiPage would be. If the WikiNodeId doesn't correspond to any WikiNode, the test will fail (and this is correct). So, we need to override the "random" creation to set a WikiNodeId of a WikiNode that exists. In the last commit can see the changes for this.

### CODE REVIEW

File: modules/util/portal-tools-rest-builder/src/main/resources/com/liferay/portal/tools/rest/builder/dependencies/base_resource_test_case.ftl [hunk 1]
```diff
@@ -1272,8 +1272,58 @@ public abstract class Base${schemaName}ResourceTestCase {
 						assertValid(get${schemaName}, multipartFiles);
 					</#if>
 				</#if>
+				<#if javaMethodSignature.methodName?cap_first?ends_with("ByExternalReferenceCode") >
```
Comment (javierdearcos) [inline modules/util/portal-tools-rest-builder/src/main/resources/com/liferay/portal/tools/rest/builder/dependencies/base_resource_test_case.ftl:1275]: `cap_first?`it is not needed here

**Review by javierdearcos — APPROVED** [review]

**Comment by javierdearcos** [comment]
I think the best strategy for Account is to override the test for the moment, and only test the update. Then we can communicate with them and send them a PR to include the insert and make their case consistent with the rest of the Portal

File: modules/util/portal-tools-rest-builder/src/main/resources/com/liferay/portal/tools/rest/builder/dependencies/base_resource_test_case.ftl [hunk 2]
```diff
@@ -1272,8 +1272,58 @@ public abstract class Base${schemaName}ResourceTestCase {
...
+					${schemaName} new${schemaName} = test${javaMethodSignature.methodName?cap_first}_create${schemaName}();
+
+                    put${schemaName} = ${schemaVarName}Resource.${javaMethodSignature.methodName}(
+
+                    <#list javaMethodSignature.javaMethodParameters as javaMethodParameter>
+                        <#if freeMarkerTool.isPathParameter(javaMethodParameter, javaMethodSignature.operation) && stringUtil.equals(javaMethodParameter.parameterName, schemaVarName + "Id")>
```
Comment (nhpatt) [inline modules/util/portal-tools-rest-builder/src/main/resources/com/liferay/portal/tools/rest/builder/dependencies/base_resource_test_case.ftl]: Is this duplicated somehow? Can we extract it to a macro if it is?

File: modules/util/portal-tools-rest-builder/src/main/resources/com/liferay/portal/tools/rest/builder/dependencies/base_resource_test_case.ftl [hunk 3]
```diff
@@ -1272,8 +1272,58 @@ public abstract class Base${schemaName}ResourceTestCase {
...
+							new${schemaName}.get${javaMethodParameter.parameterName?cap_first}()
+                        <#elseif stringUtil.equals(javaMethodParameter.parameterName, "multipartBody") || stringUtil.equals(javaMethodParameter.parameterName, schemaVarName)>
+							new${schemaName}
+                        <#else>
+							null
+                        </#if>
+                        <#sep>, </#sep>
+                    </#list>
+
+                    <#if freeMarkerTool.hasRequestBodyMediaType(javaMethodSignature, "multipart/form-data")>
```
Comment (nhpatt) [inline modules/util/portal-tools-rest-builder/src/main/resources/com/liferay/portal/tools/rest/builder/dependencies/base_resource_test_case.ftl]: Is this possible? Do we have methods with ERC+ multipart?

File: modules/util/portal-tools-rest-builder/src/main/resources/com/liferay/portal/tools/rest/builder/dependencies/base_resource_test_case.ftl [hunk 4]
```diff
@@ -1272,8 +1272,58 @@ public abstract class Base${schemaName}ResourceTestCase {
...
+						, getMultipartFiles()
+                    </#if>
+
+					);
+
+					assertEquals(new${schemaName}, put${schemaName});
+					assertValid(put${schemaName});
+
+                    get${schemaName} = ${schemaVarName}Resource.${javaMethodSignature.methodName?replace("put", "get")}(
+                    <#list javaMethodSignature.pathJavaMethodParameters as javaMethodParameter>
```
Comment (nhpatt) [inline modules/util/portal-tools-rest-builder/src/main/resources/com/liferay/portal/tools/rest/builder/dependencies/base_resource_test_case.ftl]: All this indentation looks fishy in github but it's Ok, I don't know how to indent freemarker files either.

File: modules/util/portal-tools-rest-builder/src/main/resources/com/liferay/portal/tools/rest/builder/dependencies/base_resource_test_case.ftl (hunk 3 above)
Comment (LuismiBarcos) [reply modules/util/portal-tools-rest-builder/src/main/resources/com/liferay/portal/tools/rest/builder/dependencies/base_resource_test_case.ftl]: Yep, we can create or update Documents with the ERC

**Comment by LuismiBarcos** [comment]
I've extracted some code to macros and rebase with master.

**Comment by LuismiBarcos** [comment]
ci:forward

**Comment by nhpatt** [comment]
I think the best strategy for Account is to override the test for the moment, and only test the update. Then we can communicate with them and send them a PR to include the insert and make their case consistent with the rest of the Portal

> They are changing their API that does upsert to use PUT as the rest of the portal. I would ignore the test for now and send the PR

**Comment by nhpatt** [comment]
ci:test:relevant

**Comment by nhpatt** [comment]
ci:test:relevant

**Comment by nhpatt** [comment]
@javierdearcos test failure looks real, about a different behaviour in structured content APIs, can you check it?

**Comment by javierdearcos** [comment]
Sure. I'll check it and take care of the PR from now on

**Comment by javierdearcos** [comment]
ci:forward

**Pull request closed without merging**